src/QuadForm/indefiniteLLL.jl

```diff
@@ -0,0 +1,357 @@
+end
+
+################################################################################
+#                           Quadratic Forms Reduction
+################################################################################
+
+@doc Markdown.doc"""
+    quadratic_form_lll_gram_indef(G::MatElem{fmpz}, base=0) 
+                                              -> Tuple{MatElem{fmpz}, MatElem{fmpz}}
+
+Given a Gram matrix `G` of an indefinite quadratic lattice with $det(G) != 0$.
+Return the LLL-reduction of `G` or find an isotropic vector.
+"""
+function quad_form_lll_gram_indef(G::MatElem{fmpz}, base::Bool = false)
+  n = ncols(G)
+  M = one(parent(G))
+  QD = G
+  MS = identity_matrix(QQ,n)
+
+  # GSO breaks off if one of the basis vectors is isotropic
+  for i = 1:n-1
+    if(QD[i,i] == 0)
+      return _quad_form_solve_triv(G,base)
+    end
+
+    M1 = one(MS)
+    for j = 1:n
+      M1[i,j] = - QD[i,j]//QD[i,i]
+    end
+
+    M1[i,i] = 1
+    M = M*M1
+    QD = transpose(M1)*QD*M1
+  end
+
+  M = inv(M)
+  QD = transpose(M)*_abs_matrix(QD)*M
+  QD = QD*denominator(QD)
+  QD_cont = divexact(QD,content(QD))
+  QD_cont = change_base_ring(base_ring(G),QD_cont)
+  rank_QD = rank(QD_cont)
+
+  S = transpose(lll_gram_with_transform(QD_cont)[2])
+  S = S[:,ncols(S)-rank_QD+1:ncols(S)]
+
+  if(ncols(S) < n)
+    S = _complete_to_basis(S)
+  end
+
+  red = _quad_form_solve_triv(transpose(S)*G*S,base)
+
+  if(typeof(red) <: MatElem)
+    r1 = S*red
+    return r1
+  end
+
+  r1 = red[1]
+  r2 = S*red[2]
+
+  if(length(red) == 3)
+    r3 = S*red[3]
+    return r1,r2,r3
+  end
+
+  return r1,r2
+end
+
+
+@doc Markdown.doc"""
+    quad_form_lll_gram_indefgoon(G::MatElem{fmpz}, check::Bool = false) 
+                                              -> Tuple{MatElem{fmpz}, MatElem{fmpz}}
+
+Perform the LLL-reduction of the indefinite Gram matrix `G` which goes on even if an 
+isotropic vector is found. If check = true, the function checks if `G` is indeed indefinite,
+symmetric and $det(G) != 0$. 
+"""
+function quad_form_lll_gram_indefgoon(G::MatElem{fmpz}, check::Bool = false)
+
+  if(check == true)
+    if(issymmetric(G) == false || det(G) == 0 || _isindefinite_gram_matrix(change_base_ring(QQ,G)) == false)
+      error("Input should be a symmetric, invertible, indefinite matrix.")
+    end
+  end
+  red = quad_form_lll_gram_indef(G,true)
+  
+  #If no isotropic vector is found
+  if (length(red) == 2)
+    return red
+  end
+
+  U1 = red[2]
+  G2 = red[1]
+  U2 = _mathnf(G2[1,:])[2]
+  G3 = transpose(U2)*G2*U2
+  
+  #The first line of the matrix G3 only contains 0, except some 'g' on the right, where g² | det G.
+  n = ncols(G)
+  U3 = one(parent(G))
+  U3[1,n] = round(- G3[n,n]//2*1//G3[1,n])
+  G4 = transpose(U3)*G3*U3
+
+  #The coeff G4[n,n] is reduced modulo 2g
+  U = G4[[1,n],[1,n]]
+
+  if (n == 2)
+    V = zero(parent(G[:,1]))
+  else
+    V = _vecextract(G4, [1,n] , 1<<(n-1)-2)
+  end
+
+  B = _round_matrix(-inv(change_base_ring(QQ,U))*V)
```

Comment:
```suggestion
  B = -inv(change_base_ring(QQ,U))*V
  B = map_entries(round, B)
```
It is clearer and goes with what I told you about the first function